controller/src/main/java/org/jboss/as/controller/CapabilityRegistry.java

```diff
@@ -940,16 +940,33 @@ private static Set<RuntimeRequirementRegistration> findInconsistent(Map<Capabili
 
     private RuntimeCapabilityRegistration getCapabilityRegistration(String capabilityName, CapabilityScope capabilityScope) {
         SatisfactoryCapability satisfactoryCapability = findSatisfactoryCapability(capabilityName, capabilityScope, false);
+
         if (satisfactoryCapability == null) {
             if (forServer) {
-                throw ControllerLogger.MGMT_OP_LOGGER.unknownCapability(capabilityName);
+                if (isInCapability(capabilityName)) {
+                    throw ControllerLogger.MGMT_OP_LOGGER.cannotGetCapability(capabilityName);
+                } else {
+                    throw ControllerLogger.MGMT_OP_LOGGER.missingCapability(capabilityName);
+                }
+            }
+            if (isInCapability(capabilityName)) {
+                throw ControllerLogger.MGMT_OP_LOGGER.cannotGetCapabilityInContext(capabilityScope.getName(), capabilityName);
             } else {
-                throw ControllerLogger.MGMT_OP_LOGGER.unknownCapabilityInContext(capabilityName, capabilityScope.getName());
+                throw ControllerLogger.MGMT_OP_LOGGER.missingCapabilityInContext(capabilityScope.getName(), capabilityName);
             }
```

Comment:
This if / else is the area where the current information can be improved.

If findSatisfactoryCapability returns null, either there is a CapabilityId in capabilities with the requested name, or there isn't.

If there isn't, 'throw ControllerLogger.MGMT_OP_LOGGER.unknownCapability(capabilityName);' is a better message than the existing 'unknownCapabilityInContext' message. I think it's better than your 'missingCapabilityInContext' as well. Because it's simple.

If there is a CapabilityId with the requested name, then the problem is the cap is registered in a part of the model that the dependent is not allowed to access. So we can improve this message. The whole 'capabilityScope' thing is an internal detail that's not really understandable to the user. So we can provide some info that is understandable.  What's more understandable are the RegistrationPoint instances that are inside the RuntimeCapabilityRegistration values in the capabilities map, as those are basically just resource addresses.

If instead of a 'isInCapabilities' that returns a boolean, you instead had a getRegistrationPoints that returns a Set<RegistrationPoint>, then if the set isn't empty you use its content to construct an exception message that includes those points.

```
@Message(id = ###, value = "There is no satisfactory capability '%s' available to resources with capability scope '%s'. This capability is registered at address(es) '%s', and are not accessible to resources with scope '%s'.")
    IllegalStateException noSatisfactoryCapability(String capability, String scopeName, Set<RegistrationPoint> addresses, String scopeNameAgain);
```

Reply:
Thank you very much for your feedback. I force-pushed the changes you suggest.
Also, if I may ask regarding the WFCORE-785 comments, I want to ask if it's okay for these changes to stay linked with this issue or if I should create a new one.